cardano-crypto-class/src/Cardano/Crypto/EllipticCurve/BLS12_381/Internal.hs

```diff
@@ -0,0 +1,883 @@
+
+compress :: forall curve. (BLS_P curve, BLS_Curve curve) => P curve -> ByteString
+compress p = unsafePerformIO $ do
+  withP p $ \pp -> do
+    cstr <- mallocForeignPtrBytes (compressedSizeP (Proxy @curve))
+    withForeignPtr cstr $ \cstrp -> do
+      c_blst_compress cstrp pp
+      BS.packCStringLen (cstrp, compressedSizeP (Proxy @curve))
+
+serialize :: forall curve. (BLS_P curve, BLS_Curve curve) => P curve -> ByteString
+serialize p = unsafePerformIO $ do
+  withP p $ \pp -> do
+    cstr <- mallocForeignPtrBytes (serializedSizeP (Proxy @curve))
+    withForeignPtr cstr $ \cstrp -> do
+      c_blst_serialize cstrp pp
+      BS.packCStringLen (cstrp, serializedSizeP (Proxy @curve))
+
+-- | @hash msg mDST mAug@ generates the elliptic curve hash for the given
+-- message @msg@; @mDST@ and @mAug@ are the optional @aug@ and @dst@
+-- arguments.
+hash :: (BLS_P curve, BLS_Curve curve) => ByteString -> Maybe ByteString -> Maybe ByteString -> P curve
+hash msg mDST mAug = unsafePerformIO $
+  BS.useAsCStringLen msg $ \(msgPtr, msgLen) ->
+    withMaybeCStringLen mDST $ \(dstPtr, dstLen) ->
+      withMaybeCStringLen mAug $ \(augPtr, augLen) ->
+        withNewP' $ \pPtr ->
+          c_blst_hash pPtr msgPtr (fromIntegral msgLen) dstPtr (fromIntegral dstLen) augPtr (fromIntegral augLen)
+
+toAffine :: (BLS_P curve, BLS_Curve curve) => P curve -> Affine curve
+toAffine p = unsafePerformIO $
+  withP p $ \pp ->
+    withNewAffine' $ \affinePtr ->
+      c_blst_to_affine affinePtr pp
+
+fromAffine :: (BLS_P curve, BLS_Curve curve) => Affine curve -> P curve
+fromAffine affine = unsafePerformIO $
+  withAffine affine $ \affinePtr ->
+    withNewP' $ \pp ->
+      c_blst_from_affine pp affinePtr
+
+-- | Infinity check on curve points.
+isInf :: (BLS_Curve curve) => P curve -> Bool
+isInf p = unsafePerformIO $ withP p c_blst_p_is_inf
+
+affineInG :: (BLS_Curve curve) => Affine curve -> Bool
+affineInG affine = unsafePerformIO $
+  withAffine affine c_blst_affine_in_g
+
+generator :: (BLS_Curve curve) => P curve
+generator = unsafePFromPPtr c_blst_generator
+
+---- Scalar / Fr operations
+
+scalarFromFr :: Fr -> IO Scalar
+scalarFromFr fr =
+  withNewScalar' $ \scalarPtr ->
+    withFr fr $ \frPtr ->
+      c_blst_scalar_from_fr scalarPtr frPtr
+
+frFromScalar :: Scalar -> IO Fr
+frFromScalar scalar =
+    withNewFr' $ \frPtr ->
+      withScalar scalar $ \scalarPtr ->
+        c_blst_fr_from_scalar frPtr scalarPtr
+
+frFromCanonicalScalar :: Scalar -> IO (Maybe Fr)
+frFromCanonicalScalar scalar
+  | scalarCanonical scalar
+  = Just <$> frFromScalar scalar
+  | otherwise
+  = return Nothing
+
+scalarFromBS :: ByteString -> Either BLSTError Scalar
+scalarFromBS bs =
+  if success then
+    Right scalar
+  else
+    Left BLST_BAD_SCALAR
+  where
+    (success, scalar) = unsafePerformIO $
+      withNewScalar $ \scalarPtr ->
+        BS.useAsCStringLen bs $ \(cstr, l) ->
+          c_blst_scalar_from_be_bytes scalarPtr cstr (fromIntegral l)
```

Comment:
I was looking at the Go bindings and they check that the length of the byte string is not less than `sizeScalar` before passing to the conversion function. That being said, `blst_scalar_from_be_bytes` does seem to be able to handle longer byte strings, so maybe it's ok - I think it takes the last 32 bytes of the provided bytestring. Again perhaps a question of whether we want to protect user from having their input silently truncated.

Reply:
I changed the function so that now it checks that the length is the expected length. @jmhrpr , if you agree with the changes ([here](https://github.com/input-output-hk/cardano-base/pull/266/files#diff-c03c192225bf84bd0c9cf7d189bc02bf3e266f01e37a9f16c0d0ba0f8e8644aeR802-R810)), could you resolve this conversation?

Reply:
Code from Go bindings:

```
func (fr *Scalar) FromBEndian(arr []byte) *Scalar {
	nbytes := len(arr)
	if nbytes < BLST_SCALAR_BYTES ||
		!C.blst_scalar_from_be_bytes(fr, (*C.byte)(&arr[0]), C.size_t(nbytes)) {
		return nil
	}
	return fr
}
```

New code:
```
scalarFromBS bs = unsafePerformIO $ do
  BS.useAsCStringLen bs $ \(cstr, l) -> do
    (success, scalar) <- withNewScalar $ \scalarPtr -> do
      c_blst_scalar_from_be_bytes scalarPtr cstr (fromIntegral l)
    if (success && l == sizeScalar) then
      return $ Right scalar
    else
      return $ Left BLST_BAD_SCALAR
```

The difference is that in the Go bindings the conditional will short-circuit if the bytestring length is too small and the C function won't be called on the short bytestring, whereas in your code we will call the C function with a bytestring which is maybe too small (or at least, the Go bindings are explicitly preventing this from happening so I am just assuming it is "too small") and then check the length after. Maybe the C function can handle it OK, but perhaps it's just better to align with their bindings instead of figuring out what the C function is doing. I can take a closer look at the C function if you want.

Reply:
yep, you are absolutely right. I've changed it 👍